documentation/dataplane/dash-flow-resimulation.md

```diff
@@ -0,0 +1,213 @@
+   4. [1.4. Flow resimulation w/ flow HA](#14-flow-resimulation-w-flow-ha)
+   5. [1.5. Object model change summary for full flow resimulation](#15-object-model-change-summary-for-full-flow-resimulation)
+2. [2. Policy-based flow resimulation](#2-policy-based-flow-resimulation)
+   1. [2.1. Flow tracking key and pending resimulation bit](#21-flow-tracking-key-and-pending-resimulation-bit)
+   2. [2.2. Active flow tracking](#22-active-flow-tracking)
+   3. [2.3. Passive flow tracking](#23-passive-flow-tracking)
+   4. [2.4. Flow tracking key in flow HA](#24-flow-tracking-key-in-flow-ha)
+   5. [2.5. Object model change summary for policy-based flow resimulation](#25-object-model-change-summary-for-policy-based-flow-resimulation)
+3. [3. Learning-based flow resimulation](#3-learning-based-flow-resimulation)
+4. [4. On-demand flow resimulation](#4-on-demand-flow-resimulation)
+5. [5. Explicit per flow consistency (PCC) support](#5-explicit-per-flow-consistency-pcc-support)
+
+## 1. Full flow resimulation
+
+Full flow resimulation is the most common flow resimulation. For example, after ACL is updated, we will need to resimulate all flows for this pipeline, e.g., ENI.
+
+### 1.1. Flow incarnation id
+
+To implement this, A flow incarnation id is introduced:
+
+- Each pipeline will store its own flow incarnation id, which starts from 0 and changes whenever all flows needs to resimulated by updating the SAI attribute on the pipeline.
+- Each flow also stores the current id when it is created.
+
+### 1.2. Flow resimulation process
+
+With flow incarnation id, we can implement full flow resimulation as follows:
+
+- When a new packet arrives, the current flow incarnation id will be stored into the metadata bus.
+- In Conntrack Lookup stage, if we found the id stored in its flow doesn't match the one in the pipeline, we treat this flow as resimulated.
+- For simulated flows, we continue to run the later ACL and match stages instead of directly applying the actions saved in the flow.
+- In Conntrack Update stage, a new pair of flow will be generated and replaces the current flow, which finishes the process of flow resimulation.
+
+After this, the later packet will directly hit the new flow and bypass the later stages.
+
+### 1.3. Flow incarnation id overflow
+
+Some flows can be very low volume and causing flow incarnation id stops working. Say, if the flow incarnation id is 8 bits, and not it is set to 1, a low volume flow is created. And after 256 flow resimulation calls, the next packet finally arrives and see the same id, which bypasses the flow resimulation process.
+
+To solve this, whenever the incarnation id overflows, we will need to treat all flows as resimulated. Implementation-wise, this can be done by adding another bit to indicate overflow happened, then reset the flow incarnation id to 0 in each flow during the next flow aging process. After flow aging is done, we can reset the overflow bit.
```

Comment:
How would the user know when to clear the overflow flag?

Reply:
the overflow flag will be cleaned up by data plane app after flow aging process is completed. it mostly works as full flow resimulation requested flag. maybe we should just use that instead of introducing a new flag.

ideally, we should use increase on these attributes, but SAI doesn't support this operation.

Reply:
actually, rename the flag should make things more clear. will update the doc.

Reply:
the flag is renamed also explain how the workflow works.

Reply:
Then how does a user know the flow resimulation finished?

Reply:
yea, i guess we will also need to add a callback on the switch then. otherwise, the user won't be able to know indeed.

Reply:
although, what would be the granularity for reporting. and how to identify which request we are working... I need to think more on that.